cli/src/merge_tools/internal.rs

```diff
@@ -0,0 +1,610 @@
+use std::borrow::Cow;
+use std::path::Path;
+use std::sync::Arc;
+
+use itertools::Itertools;
+use jj_lib::backend::{BackendError, FileId, ObjectId, SymlinkId, TreeId, TreeValue};
+use jj_lib::diff::{find_line_ranges, Diff, DiffHunk};
+use jj_lib::matchers::EverythingMatcher;
+use jj_lib::repo_path::RepoPath;
+use jj_lib::store::Store;
+use jj_lib::tree::Tree;
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum InternalToolError {
+    #[error("Failed to record changes: {0}")]
+    Record(#[from] scm_record::RecordError),
+    #[error("Failed to read file {path:?} with ID {id:?}: {source:?}")]
+    ReadFileBackend {
+        path: RepoPath,
+        id: FileId,
+        source: BackendError,
+    },
+    #[error("Failed to read file {path:?} with ID {id:?}: {source}")]
+    ReadFileIo {
+        path: RepoPath,
+        id: FileId,
+        source: std::io::Error,
+    },
+    #[error("Failed to read symlink {path:?} with ID {id:?} from backend: {source:?}")]
+    ReadSymlink {
+        path: RepoPath,
+        id: SymlinkId,
+        source: BackendError,
+    },
+    #[error("Failed to decode UTF-8 text for item {item} (this should not happen): {source}")]
+    DecodeUtf8 {
+        source: std::str::Utf8Error,
+        item: &'static str,
+    },
+    #[error("Rendering {item} {id} is unimplemented for the builtin difftool/mergetool")]
+    Unimplemented { item: &'static str, id: String },
+    #[error("Backend error: {0:?}")]
+    BackendError(#[from] jj_lib::backend::BackendError),
+}
+
+#[derive(Clone, Debug)]
+enum FileContents {
+    Absent,
+    Text {
+        contents: String,
+        hash: Option<String>,
+        num_bytes: u64,
+    },
+    Binary {
+        hash: Option<String>,
+        num_bytes: u64,
+    },
+}
+
+/// Information about a file that was read from disk. Note that the file may not
+/// have existed, in which case its contents will be marked as absent.
+#[derive(Clone, Debug)]
+pub struct FileInfo {
+    file_mode: scm_record::FileMode,
+    contents: FileContents,
+}
+
+/// File modes according to the Git file mode conventions. used for display
+/// purposes and equality comparison.
+///
+/// TODO: let `scm-record` accept strings instead of numbers for file modes? Or
+/// figure out some other way to represent file mode changes in a jj-compatible
+/// manner?
+mod mode {
+    pub const NORMAL: usize = 0o100644;
+    pub const EXECUTABLE: usize = 0o100755;
+    pub const SYMLINK: usize = 0o120000;
+}
+
+fn describe_binary(hash: Option<&str>, num_bytes: u64) -> String {
+    match hash {
+        Some(hash) => {
+            format!("{hash} ({num_bytes}B)")
+        }
+        None => format!("({num_bytes}B)"),
+    }
+}
+
+fn buf_to_file_contents(hash: Option<String>, buf: Vec<u8>) -> FileContents {
+    let num_bytes: u64 = buf.len().try_into().unwrap();
```

Comment:
What do you think about making `num_bytes` a `usize` in the `FileContents` variants so we don't have to convert here?